### PR TITLE
Fix: 16098 main composer is focused again when user enters key if the emoji picker is not open

### PR DESCRIPTION
**Pull request by tienifr**
### Details

We should automatically focus on the Composer if we press on regular characters like a, b, c, d, ... (not non-character keys like Enter, Shift) when we're not typing on any input or textarea.

### Fixed Issues

$ https://github.com/Expensify/App/issues/16098
PROPOSAL: https://github.com/Expensify/App/issues/16098#issuecomment-1485325162

### Tests

1. Go to any chats
2. Click on any message
3. Press any character keys like a,b,c,d (not non-character keys like Enter, Shift) 
4. Verify that the composer is focused again and the above character is entered in the composer
5. Open emoji picker 
6. Press any character keys like a,b,c,d 
7. Verify that the emoji composer is focused again and the above character is entered in this composer. And the main composer is not affected

- [ ] Verify that no errors appear in the JS console

### QA Steps

1. Go to any chats
2. Click on any message
3. Press any character keys like a,b,c,d (not non-character keys like Enter, Shift) 
4. Verify that the composer is focused again and the above character is entered in the composer
5. Open emoji picker 
6. Press any character keys like a,b,c,d 
7. Verify that the emoji composer is focused again and the above character is entered in this composer. And the main composer is not affected

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/113963320/233052933-a58b468d-293d-417e-a1c9-5a093b8d01c5.mp4

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/113963320/230656324-fb443574-697d-47b1-99ea-d7ac319f2ea5.mp4

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/113963320/230656303-1a0cd2b2-7ae6-46af-8869-41c4f3be5f66.mp4

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/113963320/230656644-7dd4d41e-5123-4cc4-bcff-ae1e61f503f1.mov

</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/113963320/230657543-79dbff62-e0c3-4bb1-933c-c296b99f7006.mp4

</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/113963320/230659268-9e2b2969-22e2-4454-9b66-79631be61369.mov

</details>

### CODE REVIEW

**Comment by cead22** [comment]
Can we update this PR's title to something that summarizes what's being done in it?

**Comment by tienifr** [comment]
@rushatgabhane @PauloGasparSv any updates?

**Comment by PauloGasparSv** [comment]
Sorry for the delay here guys!

> I don't think we should implement this feature. It's a UX polish but is it even required?
> @PauloGasparSv what do you think?

I agree and I don't think we shouldn't implement this again at the moment : / 

But I don't think we should dismiss this feature just yet! **I'll bring this discussion over to slack in #expensify-open-source so more people can chime in and give their opinion on whether we should try to implement this again or not**.

**Comment by PauloGasparSv** [comment]
>  we can close this PR, thank you!

Hey @rushatgabhane I'm confused! Aren't we going to [test this](https://github.com/Expensify/App/issues/16098#issuecomment-1532127382)?

**Comment by PauloGasparSv** [comment]
@tienifr can you fix the new conflict?

Also, what do you and @rushatgabhane think of using the following test steps? I tried to add all the regressions we had so far as tests here:

### Typing should focus on the composer
1. Open any chat
2. Click on a message
3. Start typing alpha-numeric characters (like a,b,c,d… but not Enter, Shift…)
4. Verify that the composer is focused and the characters you type are inserted there

### Typing in other inputs and modals
(Coming from [here](https://github.com/Expensify/App/issues/17478))
1.  Open any chat
2. Open the Global Search and make sure the search is working
3. Blur the global search input and start typing alpha-numeric characters (like a,b,c,d… but not Enter, Shift…)
4. Make sure the characters you typed aren’t added to the search input or the Chat composer box
5. Open your profile then display name and make sure the “First name” and “Last name” are not focused then start typing alpha-numeric characters (like a,b,c,d… but not Enter, Shift…)
6. Go back to the chat and make sure those characters were not added to the composer box
7. Go to Settings > Workspaces
8. Open any workspace and go to the “Manage Members” page
9. With no input focused, start typing alpha-numeric characters (like a,b,c,d… but not Enter, Shift…) and make sure they are not added to the Chat composer box
10. Click on “Invite” and repeat the previous step

### The composer Emoji Picker
(Coming from [here](https://github.com/Expensify/App/issues/17413))
1. Open any chat
2. Click on the Emoji Picker icon in the right side of the composer box
3. Make sure you can search inside the Emoji Picker but what you type isn’t added to the composer box
4. With the Emoji Picker opened, blur the search box and start typing alpha-numeric characters (like a,b,c,d… but not Enter, Shift…)
5. Make sure the focus goes back to the Emoji Picker search box and all characters are added there instead of the Chat composer box
6. Hover any message and click on the “Add reaction” icon
7. Repeat steps 3 to 5

### Copying&Pasting - Web & Desktop only
(Coming from [here](https://github.com/Expensify/App/issues/17397) and [here](https://github.com/Expensify/App/issues/17411)	
1. Open any chat
2. Select/Highlight the text from a message
3. Copy the text using `CMD + C` or `CTRL + C`
4. Click on the composer and paste the text using  `CMD + V` or `CTRL + V`

File: src/pages/home/report/ReportActionCompose/index.js [hunk 1]
```diff
@@ -627,19 +637,60 @@ class ReportActionCompose extends React.Component {
...
+
+        this.focus();
+        this.replaceSelectionWithInput(e.key);
+    }
+
+    /**
+     * @param {String} text
+     */
+    replaceSelectionWithInput(text) {
+        const newComment = this.comment.slice(0, this.state.selection.start) + text + this.comment.slice(this.state.selection.end, this.comment.length);
```
Comment (PauloGasparSv) [inline src/pages/home/report/ReportActionCompose/index.js:674]: I think we could call `ComposerUtils.insertText` here right?

**Comment by tienifr** [comment]
> can you check why the following is happening after the workspaces page is accessed? The modal is still opened but the input is going to the composer.

After checking our code, I found out that In 

https://github.com/Expensify/App/blob/6946beb6219c87df11cc8f81bb6c31ed8e732373/src/components/Modal/BaseModal.js#L69-L77

We always call Modal.setModalVisibility(false); when we press back button (WS -> Setting). That makes our checks false

```
 keydownListener(e) {
        if (
            this.state.isFocused ||
            this.state.isEmojiPickerVisible ||
            this.props.modal.isVisible || // false
            this.props.isSmallScreenWidth ||
            EmojiPickerAction.emojiPickerRef.current.state.isEmojiPickerVisible
        ) {
            return;
        }
```

It can be out of this scope, so I suggest to create another PR to fix this bug first.

**My solution:**

We should not call Modal.setModalVisibility(false) when we click back button

Change this line 

https://github.com/Expensify/App/blob/6946beb6219c87df11cc8f81bb6c31ed8e732373/src/components/Modal/BaseModal.js#L112

to 

```
onBackButtonPress={()=>{
                    this.shouldGoBack=true;
                    this.props.onClose()
                }}
```

and update these lines 

https://github.com/Expensify/App/blob/6946beb6219c87df11cc8f81bb6c31ed8e732373/src/components/Modal/BaseModal.js#L70-L72

to
```
hideModal(callHideCallback = true) {
        if (this.props.shouldSetModalVisibility && this.shouldGoBack) {
            Modal.setModalVisibility(false);
        }
```

Please tell me what you think @PauloGasparSv @rushatgabhane

**Comment by tienifr** [comment]
> let's try that but I think it'd cause regressions if other components are assuming the visibility to be false when back is pressed.

@rushatgabhane should we fix the visibility bug in other PR because it's not related to this PR?

**Comment by PauloGasparSv** [comment]
> @rushatgabhane should we fix the visibility bug in other PR because it's not related to this PR?

I think we should try here as @rushatgabhane said!

@tienifr I also think we should treat this as related to the bug since we'll introduce the problem in this P.R. : /

Can we attempt [that](https://github.com/Expensify/App/pull/17649#issuecomment-1558947680) fix you proposed?
I'm confused on when `shouldGoBack` would be set to `false` in that logic and if it would cause regressions so I think we should try and test : )

**Comment by tienifr** [comment]
@rushatgabhane After checking our App, I can't find any cases that the index === 0 but the modal is still open. Can you help find this case? Or do you have any suggestions? Thanks so much

**Comment by tienifr** [comment]
I spend much time, but I can't find the best solution, it can cause regression some where. Can you give me any advices? Thanks @rushatgabhane @PauloGasparSv

**Comment by PauloGasparSv** [comment]
Will investigate tomorrow : )

**Comment by PauloGasparSv** [comment]
Bump @rushatgabhane

**Comment by Santhosh-Sellavel** [comment]
@tienifr Please resolve conflicts.

**Comment by Santhosh-Sellavel** [comment]
Let's close this PR @tienifr thanks!

**Pull request closed without merging**